pi/imu.py

```diff
@@ -1,59 +1,89 @@
+# SPDX-FileCopyrightText: 2020 Bryan Siepert, written for Adafruit Industries
+#
+# SPDX-License-Identifier: Unlicense
 import time
 import board
 import busio
-import adafruit_bno08x
-from adafruit_bno08x.i2c import BNO08X_I2C
 import json
 import socket
+
 from adafruit_bno08x import (
-    BNO_REPORT_GAME_ROTATION_VECTOR,
-    BNO_REPORT_LINEAR_ACCELERATION
+    BNO_REPORT_ACCELEROMETER,
+    BNO_REPORT_GYROSCOPE,
+    BNO_REPORT_MAGNETOMETER,
+    BNO_REPORT_ROTATION_VECTOR,
 )
+from adafruit_bno08x.i2c import BNO08X_I2C
+
+try:
+    i2c = busio.I2C(board.SCL, board.SDA, frequency=400000)
+    bno = BNO08X_I2C(i2c)
+
+    bno.enable_feature(BNO_REPORT_ACCELEROMETER)
+    bno.enable_feature(BNO_REPORT_GYROSCOPE)
+    bno.enable_feature(BNO_REPORT_MAGNETOMETER)
+    bno.enable_feature(BNO_REPORT_ROTATION_VECTOR)
+except Exception as e:
+    raise RuntimeError("Could not initialize IMU")
 
 HOST = "192.168.0.102"  # The server's hostname or IP address
 PORT = 2049  # The port used by the server
-CONTROL_LOOP_FREQ = 5  # Hz
+CONTROL_LOOP_FREQ = 100  # Hz
+
+def read_data() -> dict:
+    data = {}
+    # acceleration
+    accel_x, accel_y, accel_z = bno.acceleration  # pylint:disable=no-member
+    acceleration = {}
+    acceleration["x"] = accel_x
+    acceleration["y"] = accel_y
+    acceleration["z"] = accel_z
+    data["acceleration"] = acceleration
 
-# read acceleration and rotation
-def read_data(sensor):
-        data_dict = {}
-        accel_x, accel_y, accel_z = sensor.linear_acceleration  # pylint:disable=no-member
-        data_dict["acceleration"] = ("X: %0.6f  Y: %0.6f Z: %0.6f  m/s^2" % (accel_x, accel_y, accel_z))
+    # gyro
+    gyro_x, gyro_y, gyro_z = bno.gyro  # pylint:disable=no-member
+    gyroscope = {}
+    gyroscope["x"] = gyro_x
+    gyroscope["y"] = gyro_y
+    gyroscope["z"] = gyro_z
+    data["gyroscope"] = gyroscope
 
-        quat_i, quat_j, quat_k, quat_real = sensor.game_quaternion  # pylint:disable=no-member
-        data_dict["rotation vector"] = (
-            "I: %0.6f  J: %0.6f K: %0.6f  Real: %0.6f" % (quat_i, quat_j, quat_k, quat_real)
-        )
+    # magnetometer
+    mag_x, mag_y, mag_z = bno.magnetic  # pylint:disable=no-member
+    magnetometer = {}
+    magnetometer["x"] = mag_x
+    magnetometer["y"] = mag_y
+    magnetometer["z"] = mag_z
```

Comment:
We do the instantiating the exact same way for gyro, magnetometer and acceleration. Can we make a helper method for this to clean things up.